### PR TITLE
New scheduler local node

### PR DESCRIPTION
**Pull request by istoica**
Compared with previous merge, this PR adds:
- Support for multiple resource instances in the new scheduler.
- Handle correctly tasks waiting for their arguments, worker becoming available, and workers getting disconnected.
- Numerous fixes. 

All tests pass `test_basic.py` with the new scheduler except `test_submit_api` and `test_many_fractional_resources`. These will be fixed in a next PR.

Note: The new scheduler is dependent of removing the old scheduler codepath to reconstruct the actors. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 1]
```diff
@@ -464,16 +540,19 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+  return nodes_.size(); 
+}
 
-void ClusterResourceScheduler::ResourceMapToNodeResources(
+NodeResources ClusterResourceScheduler::ResourceMapToNodeResources(
     const std::unordered_map<std::string, double> &resource_map_total,
-    const std::unordered_map<std::string, double> &resource_map_available,
-    NodeResources *node_resources) {
-  node_resources->predefined_resources.resize(PredefinedResources_MAX);
+    const std::unordered_map<std::string, double> &resource_map_available) {
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]:
```suggestion
    const std::unordered_map<std::string, double> &resource_map_available) const {
```

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 2]
```diff
@@ -486,61 +565,76 @@ void ClusterResourceScheduler::ResourceMapToNodeResources(
...
     }
   }
+  return node_resources;
 }
 
-void ClusterResourceScheduler::ResourceMapToTaskRequest(
-    const std::unordered_map<std::string, double> &resource_map,
-    TaskRequest *task_request) {
+TaskRequest ClusterResourceScheduler::ResourceMapToTaskRequest(
+    const std::unordered_map<std::string, double> &resource_map) {
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]:
```suggestion
    const std::unordered_map<std::string, double> &resource_map) const {
```

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 3]
```diff
@@ -486,61 +565,76 @@ void ClusterResourceScheduler::ResourceMapToNodeResources(
...
+
   auto it = nodes_.find(client_id);
   if (it == nodes_.end()) {
-    return;
+    NodeResources node_resources;
+    node_resources.predefined_resources.resize(PredefinedResources_MAX);
+    client_id = string_to_int_map_.Insert(client_id_string);
+    nodes_.emplace(client_id, node_resources);
+    it = nodes_.find(client_id);
+    RAY_CHECK(it != nodes_.end());
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]:
```suggestion
    RAY_CHECK(nodes_.emplace(client_id, node_resources).second);
```

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 4]
```diff
@@ -613,9 +709,10 @@ void ClusterResourceScheduler::DeleteResource(const std::string &client_id_strin
 
 std::string ClusterResourceScheduler::DebugString(void) {
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]:
```suggestion
std::string ClusterResourceScheduler::DebugString(void) const {
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 5]
```diff
@@ -69,8 +69,18 @@ class TaskResourceInstances {
   /// The list of instances of each custom resource allocated to a task.
   absl::flat_hash_map<int64_t, std::vector<double>> custom_resources;
   bool operator==(const TaskResourceInstances &other);
+  /// For each resource of this request aggregate its instances.
+  TaskRequest ToTaskRequest();
   /// Get CPU instances only.
-  std::vector<double> GetCPUInstances() { return this->predefined_resources[CPU]; };
+  std::vector<double> GetCPUInstances() {  
+    if (this->predefined_resources.size()) {
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]:
```suggestion
    if (!this->predefined_resources.empty()) {
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 6]
```diff
@@ -69,8 +69,18 @@ class TaskResourceInstances {
   /// The list of instances of each custom resource allocated to a task.
   absl::flat_hash_map<int64_t, std::vector<double>> custom_resources;
   bool operator==(const TaskResourceInstances &other);
+  /// For each resource of this request aggregate its instances.
+  TaskRequest ToTaskRequest();
   /// Get CPU instances only.
-  std::vector<double> GetCPUInstances() { return this->predefined_resources[CPU]; };
+  std::vector<double> GetCPUInstances() {
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]:
```suggestion
  std::vector<double> GetCPUInstances() const {  
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 7]
```diff
@@ -149,6 +159,9 @@ class ClusterResourceScheduler {
       const std::string &local_node_id,
       const std::unordered_map<std::string, double> &local_node_resources);
 
+  // Mapping from predefined resource indexes to resource strings
+  std::string GetResourceNameFromIndex(int64_t res_idx);
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: This doesn't seem like it needs to be a class method. Please make it a helper function in the .cc file and do not expose it in the header file.

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 8]
```diff
@@ -69,8 +69,18 @@ class TaskResourceInstances {
   /// The list of instances of each custom resource allocated to a task.
   absl::flat_hash_map<int64_t, std::vector<double>> custom_resources;
   bool operator==(const TaskResourceInstances &other);
+  /// For each resource of this request aggregate its instances.
+  TaskRequest ToTaskRequest();
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: I noticed that the `TaskRequest` class doesn't have a comment explaining what it is and how it relates to `TaskResourceInstances`. Can you add one?

By the way, `TaskRequest` looks like it could be a struct with all `const` members (right now it's a `class` with all non-const members).

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 9]
```diff
@@ -251,15 +264,13 @@ class ClusterResourceScheduler {
   int64_t NumNodes();
 
   /// Convert a map of resources to a TaskRequest data structure.
-  void ResourceMapToTaskRequest(
-      const std::unordered_map<std::string, double> &resource_map,
-      TaskRequest *task_request);
+  TaskRequest ResourceMapToTaskRequest(
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: This doesn't seem like it needs to be a class method. Please make it a helper function in the .cc file and do not expose it in the header file.

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 10]
```diff
@@ -251,15 +264,13 @@ class ClusterResourceScheduler {
...
+      const std::unordered_map<std::string, double> &resource_map);
 
   /// Convert a map of resources to a TaskRequest data structure.
-  void ResourceMapToNodeResources(
+  NodeResources ResourceMapToNodeResources(
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: This doesn't seem like it needs to be a class method. Please make it a helper function in the .cc file and do not expose it in the header file.

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 11]
```diff
@@ -343,25 +356,46 @@ class ClusterResourceScheduler {
   ///
   /// \param available A list of available capacities for resource's instances.
   /// \param resource_instances List of the resource instances being updated.
-  void AddAvailableResourceInstances(std::vector<double> available,
-                                     ResourceInstanceCapacities *resource_instances);
+  ///
+  /// \return Overflow capacities of "resource_instances" after adding instance
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: I don't understand this comment. Can you explain what overflow capacity is? It seemed like it could be either that more resources than the total capacity were marked as available, or it's the new value for (capacity - available).

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 12]
```diff
@@ -343,25 +356,46 @@ class ClusterResourceScheduler {
...
+  std::vector<double> AddAvailableResourceInstances(std::vector<double> available,
+                                                    ResourceInstanceCapacities *resource_instances);
 
   /// Decrease the available capacities of the instances of a given resource.
   ///
   /// \param free A list of capacities for resource's instances to be freed.
   /// \param resource_instances List of the resource instances being updated.
-  void SubtractAvailableResourceInstances(std::vector<double> free,
-                                          ResourceInstanceCapacities *resource_instances);
+  /// \return Under capacities of "resource_instances" after substracting instance
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: Same comment here, can you explain what under capacities means? Giving a formula would be helpful.

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 13]
```diff
@@ -343,25 +356,46 @@ class ClusterResourceScheduler {
...
-  /// \param cpu_instances Cpu instances to be removed from available cpus.
-  void SubtractCPUResourceInstances(std::vector<double> &cpu_instances);
+  /// \param cpu_instances CPU instances to be removed from available cpus.
+  ///
+  /// \return Underflow capacities of CPU instances after subtracting CPU 
+  /// capacities in cpu_instances. 
+  std::vector<double> SubtractCPUResourceInstances(std::vector<double> &cpu_instances);
+
+  bool AllocateTaskResources(int64_t node_id, const TaskRequest &task_req, 
+                            TaskResourceInstances *task_allocation /* return */);
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]:
```suggestion
                            TaskResourceInstances *task_allocation);
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 14]
```diff
@@ -343,25 +356,46 @@ class ClusterResourceScheduler {
...
   ///
-  /// \param cpu_instances Cpu instances to be removed from available cpus.
-  void SubtractCPUResourceInstances(std::vector<double> &cpu_instances);
+  /// \param cpu_instances CPU instances to be removed from available cpus.
+  ///
+  /// \return Underflow capacities of CPU instances after subtracting CPU 
+  /// capacities in cpu_instances. 
+  std::vector<double> SubtractCPUResourceInstances(std::vector<double> &cpu_instances);
+
+  bool AllocateTaskResources(int64_t node_id, const TaskRequest &task_req,
```
Comment (stephanie-wang) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: Document. Same for below methods.

Also, do we really need 3 different methods for allocating task resources? Maybe one of these should be a private method?

File: src/ray/raylet/node_manager.cc [hunk 15]
```diff
@@ -552,9 +553,34 @@ void NodeManager::NodeRemoved(const GcsNodeInfo &node_info) {
   // node.
   rpc::Address address;
   address.set_raylet_id(node_info.node_id());
+  RAY_LOG(WARNING) << "HandleUnexpectedWorkerFailure 1.0";
   HandleUnexpectedWorkerFailure(address);
 }
 
+// XXX
+std::string NodeManager::PrintLeasedWorkers() {
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Please remove these. If you really need the print messages, pass the state that you need to check (e.g., `leased_workers_`) into an anonymous helper function instead of making it a method on the `NodeManager`.

File: src/ray/raylet/node_manager.cc [hunk 16]
```diff
@@ -616,7 +646,9 @@ void NodeManager::ResourceCreateUpdated(const ClientID &client_id,
                                                      new_resource_capacity);
     }
     if (new_scheduler_enabled_) {
-      new_resource_scheduler_->UpdateResourceCapacity(client_id.Binary(), resource_label,
+      RAY_LOG(WARNING) << "====x ResourceCreateUpdated -> UpdateResourceCapacity ====";
+      new_resource_scheduler_->UpdateResourceCapacity(client_id.Binary(),
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Why do these take in `client_id.Binary()` instead of `client_id`?

File: src/ray/raylet/node_manager.cc [hunk 17]
```diff
@@ -1589,9 +1715,9 @@ void NodeManager::HandleRequestWorkerLease(const rpc::RequestWorkerLeaseRequest
   }
 
   if (new_scheduler_enabled_) {
-    auto request_resources = task.GetTaskSpecification().GetRequiredResources();
+    auto task_specs = task.GetTaskSpecification();
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]:
```suggestion
    auto task_spec = task.GetTaskSpecification();
```

File: src/ray/raylet/node_manager.cc [hunk 18]
```diff
@@ -1600,9 +1726,48 @@ void NodeManager::HandleRequestWorkerLease(const rpc::RequestWorkerLeaseRequest
             reply->mutable_worker_address()->set_port(worker->Port());
             reply->mutable_worker_address()->set_worker_id(worker->WorkerId().Binary());
             reply->mutable_worker_address()->set_raylet_id(self_node_id_.Binary());
+            RAY_LOG(WARNING) << "leased_worker 2 = " << worker->WorkerId();
             RAY_CHECK(leased_workers_.find(worker->WorkerId()) == leased_workers_.end());
             leased_workers_[worker->WorkerId()] = worker;
-            leased_worker_resources_[worker->WorkerId()] = request_resources;
+// TODO (Ion): Fix handling floating point errors, maybe by moving to integers.               
+#define ZERO_CAPACITY 1.0e-5
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Please make this a constant, such as `const kZeroCapacity` and document what it's used for.

File: src/ray/raylet/node_manager.cc [hunk 19]
```diff
@@ -1600,9 +1726,48 @@ void NodeManager::HandleRequestWorkerLease(const rpc::RequestWorkerLeaseRequest
...
+            TaskResourceInstances allocated_resources;
+            if (task_specs.IsActorCreationTask()) {
+              allocated_resources = worker->GetLifetimeAllocatedInstances();                                                   
+            } else {
+              allocated_resources = worker->GetAllocatedInstances();                                                   
+            }
+            auto predefined_resources = allocated_resources.predefined_resources;
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]:
```suggestion
            const auto &predefined_resources = allocated_resources.predefined_resources;
```

File: src/ray/raylet/node_manager.cc [hunk 20]
```diff
@@ -1600,9 +1726,48 @@ void NodeManager::HandleRequestWorkerLease(const rpc::RequestWorkerLeaseRequest
...
+            for (size_t res_idx = 0; res_idx < predefined_resources.size(); res_idx++) {
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: All of the protobuf formatting seems like it could be a helper function that takes in the predefined resources and the reply.

File: src/ray/raylet/node_manager.cc [hunk 21]
```diff
@@ -2130,47 +2280,26 @@ void NodeManager::HandleDirectCallTaskBlocked(const std::shared_ptr<Worker> &wor
...
 
-  if (!worker || worker->GetAssignedTaskId().IsNil()) {
+  if (!worker || worker->GetAssignedTaskId().IsNil() || !worker->IsBlocked()) {
     return;  // The worker may have died or is no longer processing the task.
   }
   TaskID task_id = worker->GetAssignedTaskId();
-
-  // First, always release task dependencies. This ensures we don't leak resources even
-  // if we don't need to unblock the worker below.
-  task_dependency_manager_.UnsubscribeGetDependencies(task_id);
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Why this change?

File: src/ray/raylet/node_manager.cc [hunk 22]
```diff
@@ -2418,26 +2550,39 @@ void NodeManager::AssignTask(const std::shared_ptr<Worker> &worker, const Task &
...
   TaskID task_id = worker.GetAssignedTaskId();
   RAY_LOG(DEBUG) << "Finished task " << task_id;
+  RAY_LOG(WARNING) << "FinishAssignedTask " << task_id;
 
-  // (See design_docs/task_states.rst for the state transition diagram.)
   Task task;
-  RAY_CHECK(local_queues_.RemoveTask(task_id, &task));
+  if (new_scheduler_enabled_) {
+    task = worker.GetAssignedTask();
+    leased_workers_.erase(worker.WorkerId()); // Maybe RAY_CHECK ???
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Wasn't this already erased in `HandleReturnWorker`? I don't think we want to erase it here.

File: src/ray/raylet/node_manager.cc [hunk 23]
```diff
@@ -2763,33 +2909,44 @@ void NodeManager::HandleObjectLocal(const ObjectID &object_id) {
...
-    // their assigned task ID is the actor ID.
-    for (const auto &id : ready_task_id_set_copy) {
-      RAY_CHECK(actor_registry_.count(id.ActorId()) > 0);
-      ready_task_id_set.erase(id);
+  if (new_scheduler_enabled_) {
+    for (auto task_id: ready_task_ids) {
+      auto it = waiting_tasks_.find(task_id);
+      if (it != waiting_tasks_.end()) {
+        task_dependency_manager_.UnsubscribeGetDependencies(task_id);
+        tasks_to_dispatch_.push_back(it->second);
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Don't you need to dispatch these in case tasks got added to `tasks_to_dispatch_`?

File: src/ray/raylet/node_manager.h [hunk 24]
```diff
@@ -712,6 +709,8 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
   std::deque<std::pair<ScheduleFn, Task>> tasks_to_schedule_;
   /// Queue of lease requests that should be scheduled onto workers.
   std::deque<std::pair<ScheduleFn, Task>> tasks_to_dispatch_;
+  /// Queue storing the tasks waiting for their arguments to be transferred locally.
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.h]:
```suggestion
  /// Queue tasks waiting for arguments to be transferred locally.
```

**Review by stephanie-wang — COMMENTED** [review]
I left a bunch of comments, but I'm not confident in my understanding of this PR.

Can you:
1. Remove all the debug statements. It's hard to read the diff right now with the extra statements.
2. Add comments for the `cluster_resource_scheduler.cc`. I didn't take a close look at this code because it was a bit too much to understand at once.
3. Explain the changes in the PR description. Specifically, it would be good to know what the status was before and after these changes. Also, it would help if you could add some description to explain the differences between the "old" scheduler and the new one.

File: src/ray/raylet/node_manager.cc (hunk 23 above)
Comment (istoica) [reply src/ray/raylet/node_manager.cc]: Good catch. Thanks

File: src/ray/raylet/node_manager.cc (hunk 22 above)
Comment (istoica) [reply src/ray/raylet/node_manager.cc]: Correct. Deleted.

File: src/ray/raylet/node_manager.cc (hunk 21 above)
Comment (istoica) [reply src/ray/raylet/node_manager.cc]: No longer there.

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 7 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: Not, it needs as I am using it in the node_manager.cc

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 14 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: Good point. AllocatedTaskResources() is private now.

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 9 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: Ok. It uses string string_to_int_map_ but I'll have this as an argument.

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 11 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 12 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: done

File: src/ray/raylet/node_manager.h (hunk 24 above)
Comment (istoica) [reply src/ray/raylet/node_manager.h]: done

File: src/ray/raylet/worker.h [hunk 25]
```diff
@@ -71,11 +73,35 @@ class Worker {
...
-  /// Cpus borrowed by the worker. This happens when the machine is oversubscribed
-  /// and the worker does not get back the cpu resources when unblocked.
-  /// TODO (ion): Add methods to access this variable.
-  /// TODO (ion): Investigate a more intuitive alternative to track these Cpus.
-  ResourceSet borrowed_cpu_resources_;
+  // Setter, geter, and clear methods  for allocated_instances_.
+  void SetAllocatedInstances(TaskResourceInstances &allocated_instances) { 
+      allocated_instances_ = allocated_instances;
+  };                                                   
+  TaskResourceInstances &GetAllocatedInstances() {return allocated_instances_; };
```
Comment (edoakes) [inline src/ray/raylet/worker.h]: Add newlines between method definitions please. Same with others in this file.

File: src/ray/raylet/worker.h [hunk 26]
```diff
@@ -71,11 +73,35 @@ class Worker {
   void DirectActorCallArgWaitComplete(int64_t tag);
   void WorkerLeaseGranted(const std::string &address, int port);
 
-  /// Cpus borrowed by the worker. This happens when the machine is oversubscribed
-  /// and the worker does not get back the cpu resources when unblocked.
-  /// TODO (ion): Add methods to access this variable.
-  /// TODO (ion): Investigate a more intuitive alternative to track these Cpus.
-  ResourceSet borrowed_cpu_resources_;
+  // Setter, geter, and clear methods  for allocated_instances_.
+  void SetAllocatedInstances(TaskResourceInstances &allocated_instances) {
```
Comment (edoakes) [inline src/ray/raylet/worker.h]: Should this be a unique_ptr to avoid the copy?

File: src/ray/raylet/worker.h [hunk 27]
```diff
@@ -71,11 +73,35 @@ class Worker {
...
-  /// TODO (ion): Investigate a more intuitive alternative to track these Cpus.
-  ResourceSet borrowed_cpu_resources_;
+  // Setter, geter, and clear methods  for allocated_instances_.
+  void SetAllocatedInstances(TaskResourceInstances &allocated_instances) { 
+      allocated_instances_ = allocated_instances;
+  };                                                   
+  TaskResourceInstances &GetAllocatedInstances() {return allocated_instances_; };    
+  void ClearAllocatedInstances() {
+    TaskResourceInstances nothing;  
+    allocated_instances_ = nothing; // Clear allocated instances.
```
Comment (edoakes) [inline src/ray/raylet/worker.h]: unique_ptr would be cleaner here too

File: src/ray/raylet/worker.h [hunk 28]
```diff
@@ -71,11 +73,35 @@ class Worker {
   void DirectActorCallArgWaitComplete(int64_t tag);
   void WorkerLeaseGranted(const std::string &address, int port);
 
-  /// Cpus borrowed by the worker. This happens when the machine is oversubscribed
-  /// and the worker does not get back the cpu resources when unblocked.
-  /// TODO (ion): Add methods to access this variable.
-  /// TODO (ion): Investigate a more intuitive alternative to track these Cpus.
-  ResourceSet borrowed_cpu_resources_;
+  // Setter, geter, and clear methods  for allocated_instances_.
```
Comment (edoakes) [inline src/ray/raylet/worker.h]: These comments probably aren't necessary

File: src/ray/raylet/worker.h [hunk 29]
```diff
@@ -120,6 +146,22 @@ class Worker {
...
+  TaskResourceInstances allocated_instances_;
+  /// The capacity of each resource instance allocated to this worker 
+  /// when running as an actor. 
+  TaskResourceInstances lifetime_allocated_instances_;
+  /// CPUs borrowed by the worker. This happens in the following scenario:
+  /// 1) Worker A is blocked, so it donates its CPUs back to the node.
+  /// 2) Other workers are scheduled and are allocated some of the CPUs donated by A.
+  /// 3) Task A is unblocked, but it cannot get all CPUs back. At this point,
+  /// the node is oversubscribed. borrowed_cpu_instances_ represents the number
+  /// of CPUs this node is iversubscribed by.
```
Comment (edoakes) [inline src/ray/raylet/worker.h]:
```suggestion
  /// of CPUs this node is oversubscribed by.
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 30]
```diff
@@ -343,28 +370,66 @@ class ClusterResourceScheduler {
...
 
   /// Decrease the available capacities of the instances of a given resource.
   ///
   /// \param free A list of capacities for resource's instances to be freed.
   /// \param resource_instances List of the resource instances being updated.
-  void SubtractAvailableResourceInstances(std::vector<double> free,
-                                          ResourceInstanceCapacities *resource_instances);
+  /// \return Underflow of "resource_instances" after substracting instance 
+  /// capacities in "available", i.e.,.
+  /// max(available - reasource_instances.available, 0)
```
[the remark this answers is not in the document]
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: Yes. Say you have 
- 1 free CPU
- Task1 allocates 0.8 CPUs -> available CPU = 0.2
- Task 1 blocks and gives back its 0.8CPU -> available CPU = 1
- Task 2 starts and requires 0.7 CPUs -> available CPU = 0.3
- Task 1 unblocks, and the policy is to allow it to run by giving back its CPU -> available CPU = 0.3 - 0.8 = -0.5

File: src/ray/raylet/worker.h (hunk 25 above)
Comment (istoica) [reply src/ray/raylet/worker.h]: done

File: src/ray/raylet/worker.h (hunk 26 above)
Comment (istoica) [reply src/ray/raylet/worker.h]: done

File: src/ray/raylet/worker.h (hunk 28 above)
Comment (istoica) [reply src/ray/raylet/worker.h]: done

File: src/ray/raylet/worker.h (hunk 27 above)
Comment (edoakes) [reply src/ray/raylet/worker.h]: `unique_ptr.clear()`

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 31]
```diff
@@ -162,7 +162,7 @@ bool NodeResources::operator==(const NodeResources &other) {
   return true;
 }
 
-std::string NodeResources::DebugString() const {
+std::string NodeResources::DebugString(StringIdMap string_to_in_map) const {
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Is this supposed to be `string_to_id_map`? Otherwise not sure what it means.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 32]
```diff
@@ -885,7 +886,9 @@ bool ClusterResourceScheduler::AllocateResourceInstances(
 }
 
 bool ClusterResourceScheduler::AllocateTaskResourceInstances(
-    const TaskRequest &task_req, TaskResourceInstances *task_allocation /* return */) {
+    const TaskRequest &task_req, 
+    std::shared_ptr<TaskResourceInstances> task_allocation /* return */) {
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: By convention this should be a pointer to a `shared_ptr`. Then you can remove the `return` comment because it's clear.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 33]
```diff
@@ -1043,13 +1048,16 @@ void ClusterResourceScheduler::AllocateRemoteTaskResources(
...
-  if (!task_allocation.IsEmpty()) {
-    FreeTaskResourceInstances(task_allocation);
-    UpdateLocalAvailableResourcesFromResourceInstances();
+void ClusterResourceScheduler::FreeLocalTaskResources(std::shared_ptr<TaskResourceInstances> task_allocation) {
+  if (task_allocation == nullptr) {
+    return;
   }
+  if (task_allocation->IsEmpty()) {
+    return;
+  }
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]:
```suggestion
  if (!task_allocation || task_allocation->IsEmpty()) {
    return;
  }
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 34]
```diff
@@ -14,7 +14,8 @@
 /// List of predefined resources.
 enum PredefinedResources { CPU, MEM, GPU, TPU, PredefinedResources_MAX };
 // Specify resources that consists of unit-size instances.
-static std::unordered_set<int64_t> UnitInstanceResources{CPU, GPU, TPU};
+// static std::unordered_set<int64_t> UnitInstanceResources{CPU, GPU, TPU};
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: don't think this should be here

File: src/ray/raylet/node_manager.cc [hunk 35]
```diff
@@ -1532,28 +1532,36 @@ void NodeManager::ProcessSubmitTaskMessage(const uint8_t *message_data) {
 
 void NodeManager::DispatchScheduledTasksToWorkers() {
   RAY_CHECK(new_scheduler_enabled_);
-  size_t queue_size = tasks_to_dispatch_.size(); // XXX
-  while (queue_size > 0) { //
+  size_t queue_size = tasks_to_dispatch_.size(); 
+
+  // Check every task in task_to_dispatch queue to see
+  // whether it can be dispatched and ran. This avoids head-of-line
+  // blocking where a task which cannot be dispatches because
```
Comment (edoakes) [inline src/ray/raylet/node_manager.cc]:
```suggestion
  // blocking where a task which cannot be dispatched because
```

File: src/ray/raylet/node_manager.cc [hunk 36]
```diff
@@ -1532,28 +1532,36 @@ void NodeManager::ProcessSubmitTaskMessage(const uint8_t *message_data) {
...
+  while (queue_size > 0) { 
     auto task = tasks_to_dispatch_.front(); 
     auto reply = task.first;
     auto spec = task.second.GetTaskSpecification();
+
     tasks_to_dispatch_.pop_front(); 
     queue_size--; 
 
     std::shared_ptr<Worker> worker = worker_pool_.PopWorker(spec);
     if (worker == nullptr) {
```
Comment (edoakes) [inline src/ray/raylet/node_manager.cc]:
```suggestion
    if (!worker) {
```

File: src/ray/raylet/node_manager.cc [hunk 37]
```diff
@@ -1575,7 +1583,19 @@ void NodeManager::DispatchScheduledTasksToWorkers() {
...
-  while (!tasks_to_schedule_.empty()) {
+  size_t queue_size = tasks_to_schedule_.size();
+
+  // Check every task in task_to_schedule queue to see
+  // whether it can be scheduled. This avoids head-of-line
+  // blocking where a task which cannot be scheduled because
+  // there are not enough available resources blocks other
+  // tasks from being scheduled.
+  while (queue_size > 0) {
+    if (queue_size == 0) {
```
Comment (edoakes) [inline src/ray/raylet/node_manager.cc]: This condition is unreachable - the while loop would break

Comment (edoakes) [reply src/ray/raylet/node_manager.cc]: Could you use a for loop instead? `for (size_t queue_size = tasks_to_schedule_.size(); queue_size > 0; queue_size--)`

File: src/ray/raylet/worker.h [hunk 38]
```diff
@@ -74,33 +74,32 @@ class Worker {
...
-  void SetAllocatedInstances(TaskResourceInstances &allocated_instances) { 
+  void SetAllocatedInstances(std::shared_ptr<TaskResourceInstances> allocated_instances) { 
       allocated_instances_ = allocated_instances;
+  };                            
+
+  std::shared_ptr<TaskResourceInstances> GetAllocatedInstances() {return allocated_instances_; };    
+  
+  void ClearAllocatedInstances() { allocated_instances_ = nullptr; };
+
+  void SetLifetimeAllocatedInstances(std::shared_ptr<TaskResourceInstances> allocated_instances) {
```
Comment (edoakes) [inline src/ray/raylet/worker.h]:
```suggestion
  void SetLifetimeAllocatedInstances(std::shared_ptr<TaskResourceInstances> &allocated_instances) { 
```

This will avoid incrementing the shared pointer once, which can be expensive

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 34 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: good catch!

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 31 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Yes, this argument is the `string_id_map_` in `ClusterResourceScheduler`, and keeps the mapping between the id and the name (string) of each custom resource. Ids are used to improve the performance; it's much cheaper to compare integers than strings. 

Initially, `DebugString` was a method of `ClusterResourceScheduler` but Stephanie convinced me to move it out, so now we need to pass `string_to_id_map_` as an argument.

**Review by edoakes — APPROVED** [review]
